src/coreclr/jit/morph.cpp

```diff
@@ -7773,7 +7773,7 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
                             }
 
                             assert(lcl == rhs->AsLclVarCommon()->GetLclNum());
-                            lcl = rhs->AsLclVarCommon()->GetLclNum();
+                            lcl = asgNode->gtGetOp1()->AsLclVarCommon()->GetLclNum();
```

Comment:
This code took me quite a while to understand. I have a few of suggestions:
1. Can we rename `nextNextBlock` to `retBlock`?
2. Can we mark the parts that are debug only checks under #ifdef DEBUG? E.g. the loop above is debug only now.
3. Maybe add some comments with the shapes we allow? If I understand correctly we allow an arbitrary amount of blocks containing nops or successive assignments, so basically the allowed sequences are:
```
{ GT_ASG(t_0, GT_CALL(...)) }
{ GT_ASG(t_1, t0) } (with casts on rhs potentially)
...
{ GT_ASG(t_n, t_(n - 1)) }
{ GT_RET t_n }
...
```
? Do I understand that right?

By the way, it is unclear to me that we can allow all forms of casts here.

Reply:
This is a very late validation that the tail call is plausible based on the residual IR from the call site to the ret.

If the casts we see here shouldn't be allowed we should have have blocked the tail call earlier on.

Reply:
Right, I get that this is validation, but even the validation code is overly relaxed since certain casts would not be safe to have. Given that, how useful is this validation? It also seems like it would look simpler in LIR.

Reply:
I don't think the validation is all that useful. I'd be fine if we just deleted it -- this is just the latest in a series of tweaks to keep it from reporting false positives.

Reply:
I've addressed your feedback, feel free to remove it or move to lower, I was just wanting to unblock tests